### PR TITLE
(Old) Palmer's algorithm for graphs that meet Ore's condition 

### PR DESCRIPTION
**Pull request by AlexandruValeanu**
#496 without unnecessary commits.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 1]
```diff
@@ -479,6 +479,34 @@
...
+     * @param <E> the graph edge type
+     * @return true if the graph meets Ore's condition, false otherwise
+     * @see org.jgrapht.alg.tour.PalmerHamiltonianCycle
+     */
+    public static <V, E> boolean hasOreProperty(Graph<V, E> graph){
+        Objects.requireNonNull(graph, GRAPH_CANNOT_BE_NULL);
+
+        final int n = graph.vertexSet().size();
+
+        if (!graph.getType().isSimple() || n < 3)
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: Now I understand what you were saying. I missed this as it was not at the very beginning of the method.

**Comment by d-michail** [comment]
@jsichi @jkinable Does this fit in the "tour" or in the "cycle" package?

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java (hunk 1 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: No problem, now the docs are even better!

**Comment by jsichi** [comment]
Regarding packaging, tour seems correct to me now that we have it.  And probably HierholzerEulerianCycle should be moved to the tour package.  The cycle package should be limited to algorithms for enumerating the plain cycles of a graph.

**Comment by jkinable** [comment]
Yep I agree: I would put palmer's algorithm in the tour package.

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 2]
```diff
@@ -479,6 +479,35 @@
         return new HierholzerEulerianCycle<V, E>().isEulerian(graph);
     }
 
+    /**
+     * Tests whether an undirected graph meets Ore's condition to be Hamiltonian.
+     *
+     * Let G be a (finite and simple) graph with n ≥ 3 vertices. We denote by deg v the degree of a vertex v in G, i.e. the number of incident edges in G to v.
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: Please use math notation, see contributors guidelines in the wiki

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 3]
```diff
@@ -479,6 +479,35 @@
...
+     * Then, Ore's theorem states that if deg v + deg w ≥ n for every pair of distinct non-adjacent vertices v and w of G, then G is Hamiltonian.
+     *
+     * @param graph the input graph
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     * @return true if the graph meets Ore's condition, false otherwise
+     * @see org.jgrapht.alg.tour.PalmerHamiltonianCycle
+     */
+    public static <V, E> boolean hasOreProperty(Graph<V, E> graph){
+        Objects.requireNonNull(graph, GRAPH_CANNOT_BE_NULL);
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: can be removed; requireUndirected(graph) takes care of this

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java (hunk 2 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: Done.

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java (hunk 3 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: Done.

File: jgrapht-core/src/main/java/org/jgrapht/alg/tour/PalmerHamiltonianCycle.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+ *
+ * <p>
+ *     This is an implementation of the algorithm described by E. M. Palmer in his paper. The algorithm takes a simple
+ *     graph that meets Ore's condition (see {@link GraphTests#hasOreProperty(Graph)}) and returns a Hamiltonian cycle.
+ *     The algorithm runs in $O(|V|^2)$ and uses $O(|V|)$ space.
+ * </p>
+ *
+ * <p>
+ *    See <a href="https://en.wikipedia.org/wiki/Ore%27s_theorem">wikipedia</a> for a short description of Ore's theorem
+ *    and the resulting algorithm and <a href="https://www.sciencedirect.com/science/article/pii/S0898122197002253">Palmer's paper</a> for more details.
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/tour/PalmerHamiltonianCycle.java]: please use proper citation. If this link changes, we no longer know which paper you were referencing.

Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/tour/PalmerHamiltonianCycle.java]: Done.

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 5]
```diff
@@ -479,6 +479,36 @@
         return new HierholzerEulerianCycle<V, E>().isEulerian(graph);
     }
 
+    /**
+     * Tests whether an undirected graph meets Ore's condition to be Hamiltonian.
+     *
+     * Let G be a (finite and simple) graph with $n ≥ 3$ vertices. We denote by deg(v) the degree of a vertex v in G,
+     * i.e. the number of incident edges in G to v.
+     * Then, Ore's theorem states that if $deg(v) + deg(w) ≥ n$ for every pair of distinct non-adjacent vertices
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: please replace ≥ by \geq

**Comment by jkinable** [comment]
@d-michail I would rename the TSPAlgorithm interface by HamiltonianCycleAlgorithm (the implementation remains exactly the same, just a name change). That way, this class can also implement this interface. What do you think?

**Comment by AlexandruValeanu** [comment]
@jkinable Please note that in the description of the 'TSPAlgorithm' interface there is the description of the TSP problem: "Given a list of cities and the distances between each pair of cities, what is the **shortest** possible route that visits each city exactly once and returns to the origin city?".

This algorithm has nothing to do with costs and therefore it doesn't actually solve the TSP problem. This is the main reason why I didn't make it implement the interface.

**Comment by jkinable** [comment]
@AlexandruValeanu yes I know, that's why a name change is required. It wouldnt make sense to implement the TSPAlgorithm interface. However, every TSP is a HC, so that's fine.

**Comment by AlexandruValeanu** [comment]
@jkinable But that would also mean to change the purpose of the interface. What do you think about adding a more general interface `HamiltonianCycleAlgorithm` and make `TSPAlgorithm` extend it? This way we don't break any existing code and we get an interface that can be used by my implementation (and more that don't use costs).

**Comment by d-michail** [comment]
I would be in favor of having two different interfaces (not even one extending the other). These interfaces are mostly for namespacing. Some implementations could implement both.

**Comment by AlexandruValeanu** [comment]
Check #529 for the new version of the PR.

**Pull request closed without merging**